core/store/ledgerstore/tx_handler.go

```diff
@@ -971,10 +971,39 @@ func HandleShardCallTransaction(store store.LedgerStore, overlay *overlaydb.Over
 		case *xshard_types.XShardAbortMsg:
 			handleShardAbortMsg(msg, lockedAddress, lockedKeys, xshardDB)
 		}
+		sourceTxHash := req.GetSourceTxHash()
+		if notify.ContractEvent.SourceTxHash == nil {
+			notify.ContractEvent.SourceTxHash = make([]common.Uint256, 0)
+			notify.ContractEvent.SourceTxHash = append(notify.ContractEvent.SourceTxHash, sourceTxHash)
+		} else {
+			if !hasTxHash(sourceTxHash, notify.ContractEvent) {
+				notify.ContractEvent.SourceTxHash = append(notify.ContractEvent.SourceTxHash, sourceTxHash)
+			}
+		}
+		if notify.ContractEvent.Notify != nil {
+			for _, info := range notify.ContractEvent.Notify {
+				if info != nil {
+					info.SourceTxHash = sourceTxHash
+				}
+			}
+		}
 	}
-
 	return nil
 }
+func hasTxHash(txHash common.Uint256, notify *event.ExecuteNotify) bool {
+	if notify == nil {
```

Comment:
can put this condition check like? if  notify == nil || notify.SourceTxHash == nil || len(notify.SourceTxHash) == 0 {
        return false
    }

Reply:
I remove these check, because it has checked in where invoked hasTxHash